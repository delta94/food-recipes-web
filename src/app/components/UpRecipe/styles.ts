import styled, { css } from 'styled-components'

export const Container = styled.article`
    max-width: 700px;
    margin: 2rem auto;

    h1 {
        margin-bottom: 2rem;
    }
`

export const ContainerRecipe = styled.article`
    display: flex;

    div, form {
        width: 100%;
    }

    button {
        cursor: pointer;
        border: 0;
        border-radius: 25px;
        padding: 0.5rem 1.6rem;
        background-color: #f6e5c9;
        color: black;
    }
`

export const BoxInput = styled.div`
        display: flex;
        flex-direction: column;
        margin: 1rem;

    input, textarea {
        height: 3rem;
        border: 0;
        padding-left: 1rem;
        border-radius: 8px;
        border: 1px solid lightgrey;
        font-size: 1.1rem;
    }

    label + input, label + textarea {
        margin-top: 0.5rem;
    }

    textarea {
        resize: vertical;
        height: 6rem;
    }

`

interface IBoxImage {
    preview: string | undefined;
}

export const BoxImage = styled.div<IBoxImage>`
    ${props => props.preview && css`
        background-image: url(${props.preview});
        background-position: center;
        background-repeat: no-repeat;
        background-size: cover;
    `}

    border-radius: 8px;
    margin: 1rem;
    height: 10rem;
    background-color: white;


    label {
        width: 100%;
        height: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
    }


    .MuiSvgIcon-root {
        font-size: 3.2rem;
    }

    input {
        padding: 4rem;
        z-index: -1;
        position: absolute;
        opacity: 0;
        /* display: none; */
    }

`
