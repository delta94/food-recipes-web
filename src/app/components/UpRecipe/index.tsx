import React, { useState, useEffect, ChangeEvent } from 'react'
import { useDispatch } from 'react-redux'
import baseURL from '../../../config/url'

import { Creators } from '../../../core/redux/store/ducks/recipe'
import BackupIcon from '@material-ui/icons/Backup'

import { Container, ContainerRecipe, BoxInput, BoxImage } from './styles'

interface IRecipe {
  idFrom: string
  nameFrom: string
  imageFrom: string
  ingredientsFrom: string
  mode_prepareFrom: string
  timeFrom: string
}

const UpRecipe = ({
  idFrom,
  nameFrom,
  imageFrom,
  ingredientsFrom,
  mode_prepareFrom,
  timeFrom
}: IRecipe) => {
  const dispatch = useDispatch()

  const [selectedFile, setSelectedFile] = useState<File>()
  const [preview, setPreview] = useState<string | undefined>(() => `${baseURL}/files/${imageFrom}`)
  const [name, setName] = useState<string>(() => nameFrom)
  const [ingredients, setIngredients] = useState<string>(() => ingredientsFrom)
  const [mode_prepare, setMode_prepare] = useState<string>(() => mode_prepareFrom)
  const [time, setTime] = useState<string>(() => timeFrom)

  useEffect(() => {
    if (!selectedFile) {
      return
    }

    const objectUrl = URL.createObjectURL(selectedFile)
    setPreview(objectUrl)

    return () => URL.revokeObjectURL(objectUrl)
  }, [selectedFile])

  const handleSelectFile = (e: ChangeEvent<HTMLInputElement>) => {
    if (!e.target.files || e.target.files.length === 0) {
      setSelectedFile(undefined)
      return
    }

    setSelectedFile(e.target.files[0])
  }

  function handleSubmit (event: ChangeEvent<HTMLFormElement>) {
    event.preventDefault()

    if (selectedFile || preview) {
      dispatch(
        Creators.updateRecipeRequest(
          idFrom,
          selectedFile || imageFrom,
          name,
          ingredients,
          mode_prepare,
          time
        )
      )
    }
  }

  return (
    <Container>
      <h1>Editar sua receita</h1>

      <ContainerRecipe>
        <form onSubmit={handleSubmit}>
          <div>
            <BoxImage preview={preview}>
              <label>
                <input type='file' name='image' onChange={handleSelectFile} placeholder="Insira sua imagem" />
                <BackupIcon fontSize='large' />
              </label>
            </BoxImage>

            <BoxInput>
              <label htmlFor='name'>Nome da receita:</label>
              <input
                name='name'
                placeholder='Insira o nome da sua receita'
                value={name}
                onChange={(e) => setName(e.target.value)}
              />
            </BoxInput>

            <BoxInput>
              <label htmlFor='ingredients'>Ingredientes:</label>
              <textarea
                name='ingredients'
                placeholder='Insira os ingredientes'
                value={ingredients}
                onChange={(e) => setIngredients(e.target.value)}
              />
            </BoxInput>

            <BoxInput>
              <label htmlFor='mode_prepare'>Modo de preparo:</label>
              <textarea
                name='mode_prepare'
                placeholder='Insira o modo de preparo'
                value={mode_prepare}
                onChange={(e) => setMode_prepare(e.target.value)}
              />
            </BoxInput>

            <BoxInput>
              <label htmlFor='time'>Tempo:</label>
              <input
                name='time'
                placeholder='Insira o tempo estimado'
                value={time}
                onChange={(e) => setTime(e.target.value)}
              />
            </BoxInput>

            <button type='submit'>Atualizar receita</button>
          </div>
        </form>
      </ContainerRecipe>
    </Container>
  )
}

export default UpRecipe
