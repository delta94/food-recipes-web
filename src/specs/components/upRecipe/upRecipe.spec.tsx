import React from 'react'
import { fireEvent, getByText, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import UpRecipe from '../../../app/components/UpRecipe'

import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'

const mockStore = configureStore([])

describe('[COMPONENTS] UpRecipe', () => {
  let store: any

  beforeEach(() => {
    store = mockStore({
      auth: {}
    })
  })

  it('Should render correctly', () => {
    const container = render(
      <BrowserRouter>
        <Provider store={store}>
          <UpRecipe
            idFrom='any_id'
            nameFrom='any_name'
            imageFrom='any_image'
            ingredientsFrom='any_ingredients'
            mode_prepareFrom='any_mode_prepare'
            timeFrom='any_time'
          />
        </Provider>
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('Should return if onSubmit was called successfully', () => {
    const { getByText } = render(
      <BrowserRouter>
        <Provider store={store}>
          <UpRecipe
            idFrom='any_id'
            nameFrom='any_name'
            imageFrom='any_image'
            ingredientsFrom='any_ingredients'
            mode_prepareFrom='any_mode_prepare'
            timeFrom='any_time'
          />
        </Provider>
      </BrowserRouter>
    )

    expect(fireEvent.click(getByText('Atualizar receita'))).toBeTruthy()
  })

  // it('Should return field value', () => {
  //   const { getByPlaceholderText, container, getByLabelText } = render(
  //     <BrowserRouter>
  //       <Provider store={store}>
  //         <UpRecipe
  //           idFrom='any_id'
  //           nameFrom='any_name'
  //           imageFrom='any_image'
  //           ingredientsFrom='any_ingredients'
  //           mode_prepareFrom='any_mode_prepare'
  //           timeFrom='any_time'
  //         />
  //       </Provider>
  //     </BrowserRouter>
  //   )

  //   // fireEvent.change(getByPlaceholderText('Insira o nome da sua receita'), {
  //   //   target: { value: 'any_name' }
  //   // })

  //   // expect(getByPlaceholderText('Insira o nome da sua receita')).toHaveValue('any_name')

  //   fireEvent.change(getByLabelText(''), {
  //     target: {
  //       files: [new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png' })]
  //     }
  //   })

  //   // container.findByLabelText('input').simulate('change', { target: { value: 'matched' } })
  // })
})
