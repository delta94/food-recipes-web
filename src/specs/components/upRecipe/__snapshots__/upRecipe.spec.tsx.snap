// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENTS] UpRecipe Should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="sc-bdnylx fxXoRh"
      >
        <h1>
          Editar sua receita
        </h1>
        <article
          class="sc-gtssRu bWzDNG"
        >
          <form>
            <div>
              <div
                class="sc-hKFyIo hVTBOP"
              >
                <label>
                  <input
                    name="image"
                    type="file"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
                    />
                  </svg>
                </label>
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="name"
                >
                  Nome da receita:
                </label>
                <input
                  name="name"
                  placeholder="Insira o nome da sua receita"
                  value="any_name"
                />
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="ingredients"
                >
                  Ingredientes:
                </label>
                <textarea
                  name="ingredients"
                  placeholder="Insira os ingredientes"
                >
                  any_ingredients
                </textarea>
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="mode_prepare"
                >
                  Modo de preparo:
                </label>
                <textarea
                  name="mode_prepare"
                  placeholder="Insira o modo de preparo"
                >
                  any_mode_prepare
                </textarea>
              </div>
              <div
                class="sc-dlnjPT ksZKJP"
              >
                <label
                  for="time"
                >
                  Tempo:
                </label>
                <input
                  name="time"
                  placeholder="Insira o tempo estimado"
                  value="any_time"
                />
              </div>
              <button
                type="submit"
              >
                Atualizar receita
              </button>
            </div>
          </form>
        </article>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="sc-bdnylx fxXoRh"
    >
      <h1>
        Editar sua receita
      </h1>
      <article
        class="sc-gtssRu bWzDNG"
      >
        <form>
          <div>
            <div
              class="sc-hKFyIo hVTBOP"
            >
              <label>
                <input
                  name="image"
                  type="file"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
                  />
                </svg>
              </label>
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="name"
              >
                Nome da receita:
              </label>
              <input
                name="name"
                placeholder="Insira o nome da sua receita"
                value="any_name"
              />
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="ingredients"
              >
                Ingredientes:
              </label>
              <textarea
                name="ingredients"
                placeholder="Insira os ingredientes"
              >
                any_ingredients
              </textarea>
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="mode_prepare"
              >
                Modo de preparo:
              </label>
              <textarea
                name="mode_prepare"
                placeholder="Insira o modo de preparo"
              >
                any_mode_prepare
              </textarea>
            </div>
            <div
              class="sc-dlnjPT ksZKJP"
            >
              <label
                for="time"
              >
                Tempo:
              </label>
              <input
                name="time"
                placeholder="Insira o tempo estimado"
                value="any_time"
              />
            </div>
            <button
              type="submit"
            >
              Atualizar receita
            </button>
          </div>
        </form>
      </article>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
